### DISCUSSION TITLE
How many slip systems should be entered when calculating GND density for HCP

### DISCUSSION THREAD

**Discussion by PengWei97** [mtex-toolbox/mtex#2023] [Ask Anything] [answered]
Dear MTEX experts,

Recently, I calculated GND density based on EBSD data, but the following error was displayed during the calculation, the same as in #1813. However, he only pointed out that slip system needs to be input into `dislocationSystem.hcp()`, but not explain how many slip systems need to be input. In other words, the other two types of slip systems also need to be input in addition to the basal slip systems. How should I modify the following code so that the GND density  will not be report an error?

```bash 
Warn: There are no predefined slip systems for hcp yet. You might have to define those individually. 

In slipSystem.hcp (line 135)
In dislocationSystem.hcp (line 167)
When calling 'slipSystem. hcp', the output parameter 'sS' (which may also include other parameters) was not assigned a value.
Error locationSystem. hcp (line 167)
dS = dislocationSystem(symmetrise(slipSystem.hcp(cs),'antipodal'));
```
```bash  
% calculate GNDs
  ebsd_mesh = ebsd('indexed').gridify;
  kappa = ebsd_mesh.curvature;
  alpha = kappa.dislocationDensity; % the incomplete dislocation density tensor
  dS = dislocationSystem.hcp(ebsd_mesh.CS); % Crystallographic Dislocations
  dSRot = ebsd_mesh.orientations * dS; % to rotate the dislocation tensors into the specimen reference frame as well
```

Any suggestions or insights would be greatly appreciated. Thank you in advance for your time and assistance. I look forward to hearing from you.

Best regards,
Wei

**Comment by zmichels** [comment]
Hi,You’ll need to specify/define the slip systems you want to consider. Check out the help documentation here for syntax and examples:Slip Systems | MTEXmtex-toolbox.github.ioZachOn Dec 26, 2023, at 6:12 AM, Wei Peng ***@***.***> wrote:﻿
Dear MTEX experts,
Recently, I calculated GND density based on EBSD data, but the following error was displayed during the calculation, the same as in #1813. However, he only pointed out that slip system needs to be input into dislocationSystem.hcp(), but not explain how many slip systems need to be input. In other words, the other two types of slip systems also need to be input in addition to the basal slip systems. How should I modify the following code so that the GND density  will not be report an error?
Warn: There are no predefined slip systems for hcp yet. You might have to define those individually. 

In slipSystem.hcp (line 135)
In dislocationSystem.hcp (line 167)
When calling 'slipSystem. hcp', the output parameter 'sS' (which may also include other parameters) was not assigned a value.
Error locationSystem. hcp (line 167)
dS = dislocationSystem(symmetrise(slipSystem.hcp(cs),'antipodal'));
% calculate GNDs
  ebsd_mesh = ebsd('indexed').gridify;
  kappa = ebsd_mesh.curvature;
  alpha = kappa.dislocationDensity; % the incomplete dislocation density tensor
  dS = dislocationSystem.hcp(ebsd_mesh.CS); % Crystallographic Dislocations
  dSRot = ebsd_mesh.orientations * dS; % to rotate the dislocation tensors into the specimen reference frame as well
Any suggestions or insights would be greatly appreciated. Thank you in advance for your time and assistance. I look forward to hearing from you.
Best regards,
Wei

**Reply by PengWei97** [reply]
Thank you for your prompt reply. Meybe I did not explain it cleary. My question is when calculating the GND density of the HCP structure, is the slip systems entered here inclusive of all slip systems (basal, cylindrical, conical) of the HCP structure, or is it sufficient to pick only a specific slip system?

This one question is perhaps more on the theoretical side of the GND density calculation. Meanwhile, when I used metx-5.8.1, the calculation of GND density on shows warnings and no errors.

**Comment by kilir** [comment, marked as answer]
Hi Wei,
in order to compute the GND, mtex tries to fit the observed curvature in your ebsd map by combinations of dislocations while minimizing a number associated with each dislocation (dislocation energy). 

The GND you will get depends on how well the dislocations/slip systems you specified, are capable in producing the observed curvature.  For example, if you do not define a dislocation/slip system which otherwise could easily produce a component of the observed curvature, other, less suitable dislocations wills be used. The result will be a much higher number of dislocations required. In some cases there will be no solution at all. 

So, in orde to get a somewhat meaningful number (debatable if this case actually exists) at all, you need to make sure that you provide all possible slipsystems / dislocations required to explain the observed curvature. Because the observed curvature stems from a) something that might have physically happened to the crystal and b) eventually noise, you should provide slipsystems/dislocations for all eventualities.

Hope this helps.

Cheers,

Rüdiger

**Reply by PengWei97** [reply]
I appreciate your detailed explanation. To ensure a meaningful calculation of GND density, it's essential to account for all possible slip systems. For instance, in the case of HCP, this entails including all basal, cylindrical, and conical slip systems. I don't know if my understanding is correct or not. Could your kindly provide further clarification or guidance on this mattering? You assistance in enhancing my understanding would be greatly appreciated.

Thank you once again for your support and expertise.

Best regards,
wei
